### ISSUE TITLE
please clarify "xs" and "ys" names in seq2seq tutorial

### ISSUE THREAD

**Issue by bionicles** [facebookresearch/ParlAI#1329]
Thank you for doing a great job in making it easier to get started.

Onboarding easier = more people cite this = more models and data = faster progress

One problem: Trying to make new models for genetic programming stuffs, but it isn't clear exactly what are the variables returned when we vectorize observations because the names are just 2 letters, xs and ys, -- how can ys be the labels if there's another "labels" ?

Would it be possible to adjust the seq2seq tutorial and code to clarify what "xs" and "ys" are exactly? 

Thanks for any help,

Bion Howard, Founder at Bitpharma.com

**Comment by bionicles** [comment]
the batch_act method invokes PaddingUtils which is deprecated

**Comment by bionicles** [comment]
Is torch_agent.match_batch the new method to use instead of PaddingUtils.map_predictions?

    def match_batch(self, batch_reply, valid_inds, output=None):
https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_agent.py

**Comment by bionicles** [comment]
We want to make one system which can do many different tasks including Parlai, OpenAI, custom biomolecular engineering stuff, so it's hard to make that a subclass of torch_agent; 

maybe we could save re-implementation:
 TorchAgent.__init__() method could accept a User Agent Module --> if that is a torch.nn module, we can just use it, if the user agent module is a python class then we could invoke the user's desired forward/compute function by string name? e.g.

def __init__(self, user_agent=None, custom_forward_name="compute"):  
  super().__init__(opt, shared)
  if user_agent = None:
     ... normal stuff ... 
  else:
     if user_agent.__class__.__name__ == "Torch Module":
         self.module = user_agent
     else:
        self.module = user_agent
        self.forward = get_attr(user_agent, custom_forward_name)      

so then when our trainer decides to run the Parlai task, it can just do

```
def run_dialog_task(nature, nurture):
     system = System(nature, nurture)
     parlai_system = TorchAgent(system)
     scores = {}
     for dialog_task_name in nurture.parlai_task_names:
            parlai_teacher = ParlaiTeacher(dialog_task)
            reply = None
            for task_number in nurture.episodes_per_task:
                  batch = parlai_teacher.get_dialog(reply)
                  reply = parlai_system.reply_to(batch)
            scores[dialog_task_name] = parlai_system.get_score_for_task(dialog_task_name)
     return scores
```

**Comment by stephenroller** [comment]
Hi @bionicles,

Apologies for delays; this all came up on holiday break and I just rediscovered it.

- There isn't a perfect correspondence between the `seq2seq` we use the most, and the `example_seq2seq` presented in the tutorial. The goal of the example is to present a more minimal example where you can see how all pieces work easily. We tend to recommend users building new agents consider inheriting from `TorchAgent` or `TorchGeneratorAgent` so that more boilerplate is done for them already. Please take a look at the newer [example seq2seq](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/example_seq2seq/example_seq2seq.py) which inherits from `TorchAgent`. The newer stuff has more explicit names for these variables.
- match_batch and similar are done because the PyTorch LSTM requires the inputs to be sorted by length. This requires extra book keeping to match input batches to actual tensors, and then back again.
- If you inherit from TorchAgent, you are required to implemented `build_model`, which creates a `nn.Module`. Custom forward names will not be implemented, but you can implement `train_step` and `eval_step` to call whatever you want.

Hope that helps!
